### DISCUSSION TITLE
How to submit a cron-workflow?

### DISCUSSION THREAD

**Discussion by scape7yu** [argoproj/argo-workflows#5760] [Q&A] [answered]
We can see submit button in the ArgoUI，but I can't find a way to submit a cron-workflow right now through api, do you have any ideas?

![图片](https://user-images.githubusercontent.com/79689551/116219287-07973500-a77e-11eb-8d53-f892765df2df.png)

I think I can update the schedule time from '00 09 * * *' to now time + 2s, then the workflow will start after 2s, after then, I will update the schedule time from now time + 2s to '00 09 * * *' back.
But I don't think it is a clever method.

**Comment by terrytangyuan** [comment, marked as answer]
The timing to trigger workflows from a CronWorkflow depends on the schedule you specify. So you should be able to specify any schedule you want and wait until the next scheduled time. Once a new workflow is triggered, you may suspend the CronWorkflow from triggering new workflows from the UI. 

Alternatively, you can manually trigger new workflows from an existing CronWorkflow resource via Argo CLI:

```
argo submit --from cronwf/my-cron-wf
```

**Reply by scape7yu** [reply]
thanks very much, after trying, I've found how to start CronWorkflow right now through api, like this:

```python
from argo.workflows.client import (ApiClient,
                                   WorkflowServiceApi,
                                   Configuration,
                                   V1alpha1WorkflowSubmitRequest)

client = ApiClient(configuration=config)
service = WorkflowServiceApi(api_client=client)

namespace = "test"
name = "cronworkflow-test-7mdmd"

service.submit_workflow(namespace=namespace,
                        body=V1alpha1WorkflowSubmitRequest(namespace=namespace,
                                                           resource_kind="CronWorkflow",
                                                           resource_name=name,
                                                           submit_options=None,
                                                           local_vars_configuration=None))
```

I thought I should use `CronWorkflowServiceApi`, in fact use `WorkflowServiceApi` will be OK.

**Comment by mayankagarwal2402** [comment]
Hi @terrytangyuan, is there a way to do the same via the Argo workflows Rest API?
I have implemented workflow and cron creation using the Rest APIs. I already have scheduled a cron to run once every day, but on some occasions, I would like to run the same workflow on demand.

There are some restrictions that I want to follow while being able to this:
1. We have restricted argo-cli access only to devops. Similar restrictions will apply to the Argo-UI to allow view-only access. On a daily basis we want our workflows to be updated or triggered only via version-controlled APIs.
2. Using Golang SDK is not an extensible option for us, as not all of our projects are written in the go, python or java for which the SDKs are available.
3. We want to avoid maintaining two duplicate specs for the CronWorkflow and on-demand Workflow respectively to this end.

**Reply by rekcahpassyla** [reply]
In 3.5 and 3.6, this works:

```
curl -X POST   -H "Authorization: Bearer $ARGO_TOKEN"   -H "Content-Type: application/json"   
-d '{
        "resourceKind": "cronwf",
        "resourceName":"myworkflow"
}' "$ARGO_URL/api/v1/workflows/$ARGO_NAMESPACE/submit"
```
